### ISSUE TITLE
Gradients are zero 

### ISSUE THREAD

**Issue by Shobaky** [tensorflow/tensorflow#52175]
I have an ANN to classify food, however it doesn't work, i kept debugging till i found that the problem is with the weights aren't updated, so i printed the grads to find out that they are always zero except for the first time

Here is the code
   `def forward(x):
  
        return tf.matmul(x,W) + b`

   `def activate(x):
        return tf.nn.softmax(forward(x))`

   `def model(x):
   
        x = flatten(x)
    
        return activate(x)`

    `def cross_entropy(y_label, y_pred):
  
         return (-tf.reduce_sum(y_label * tf.math.log(y_pred + 1.e-10)))`

    `optimizer = tf.keras.optimizers.SGD(learning_rate=0.5)`

```
  `def train_step(x, y ):
             with tf.GradientTape() as tape:
             #compute loss function
             current_loss = cross_entropy( y, model(x))
             print(current_loss)
             # compute gradient of loss 
            #(This is automatic! Even with specialized funcctions!)
            grads = tape.gradient( current_loss , [W,b])
            # Apply SGD step to our Variables W and b
            print(grads)
        
            optimizer.apply_gradients( zip( grads , [W,b] ) )   
        
           return current_loss.numpy()`
```

    `W = tf.Variable(tf.zeros([196608, 3],tf.float32))
      #Bias tensor
       b = tf.Variable(tf.zeros([3],tf.float32))

       loss_values=[]
       accuracies = []
       epochs = 10
       x_train = np.empty([50,256,256,3])
       y_train = np.empty([50,1,3])

    for i in range(epochs):
         j=0
         k = 0
        # each batch has 50 examples
        for x_train_batch, y_train_batch in train_ds:
                    if j == 0 and i == 0 :
                        for k in range(50):
                             x_train[k] = x_train_batch[k]
                             y_train[k] = y_train_batch[k]
                        
            
        
                    j+=1
                    current_loss = train_step(x_train_batch/255.0, y_train_batch)
                     if j%500 == 0: #reporting intermittent batch statistics
                         print("epoch ", str(i), "batch", str(j), "loss:", str(current_loss) ) 
            
        x_train = tf.convert_to_tensor(x_train)
        y_train = tf.convert_to_tensor(y_train)`

I also checked for the current lose (if you notice it's the first output just before the gradient) and it isn't zero as you see
OUTPUT:
`tf.Tensor(54.930614, shape=(), dtype=float32)
[<tf.Tensor: shape=(196608, 3), dtype=float32, numpy=
array([[ 2.7033854,  1.9847847, -4.6881704],
       [ 2.6230955,  1.6556438, -4.278739 ],
       [ 2.289314 ,  1.2310266, -3.5203404],
       ...,
       [ 1.6812016,  1.4880723, -3.1692739],
       [ 1.6665692,  1.2746813, -2.9412503],
       [ 1.73799  ,  0.5771449, -2.3151352]], dtype=float32)>, <tf.Tensor: shape=(3,), dtype=float32, numpy=array([ 2.6666658,  6.6666684, -9.333334 ], dtype=float32)>]
tf.Tensor(483.5429, shape=(), dtype=float32)
[<tf.Tensor: shape=(196608, 3), dtype=float32, numpy=
array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       ...,
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]], dtype=float32)>, <tf.Tensor: shape=(3,), dtype=float32, numpy=array([0., 0., 0.], dtype=float32)>]
tf.Tensor(713.80145, shape=(), dtype=float32)
[<tf.Tensor: shape=(196608, 3), dtype=float32, numpy=
array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       ...,
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]], dtype=float32)>, <tf.Tensor: shape=(3,), dtype=float32, numpy=array([0., 0., 0.], dtype=float32)>]
tf.Tensor(483.5429, shape=(), dtype=float32)
[<tf.Tensor: shape=(196608, 3), dtype=float32, numpy=
array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       ...,
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]], dtype=float32)>, <tf.Tensor: shape=(3,), dtype=float32, numpy=array([0., 0., 0.], dtype=float32)>]
tf.Tensor(483.5429, shape=(), dtype=float32)
[<tf.Tensor: shape=(196608, 3), dtype=float32, numpy=
array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       ...,
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]], dtype=float32)>, <tf.Tensor: shape=(3,), dtype=float32, numpy=array([0., 0., 0.], dtype=float32)>]`

**Comment by chunduriv** [comment]
@Shobaky, 
In order to expedite the trouble-shooting process, could you please provide a complete code and the TensorFlow version you are using?Thanks!
